app/dao/notifications_dao.py

```diff
@@ -600,18 +599,27 @@ def dao_get_notifications_by_recipient_or_reference(service_id, search_term, not
         except InvalidEmailError:
             normalised = search_term.lower()
 
-    elif notification_type == LETTER_TYPE:
-        raise InvalidRequest("Only email and SMS can use search by recipient", 400)
+    elif notification_type in {LETTER_TYPE, None}:
+        # For letters, we store the address without spaces, so we need
+        # to removes spaces from the search term to match. We also do
+        # this when a notification type isn’t provided (this will
+        # happen if a user doesn’t have permission to see the dashboard)
+        # because email addresses and phone numbers will never be stored
+        # with spaces either.
+        normalised = ''.join(search_term.split()).lower()
 
     else:
-        normalised = search_term.lower()
+        raise TypeError(
+            f'Notification type must be {EMAIL_TYPE}, {SMS_TYPE}, {LETTER_TYPE} or None'
+        )
 
     normalised = escape_special_characters(normalised)
     search_term = escape_special_characters(search_term)
 
     filters = [
         Notification.service_id == service_id,
         or_(
+            Notification.to.ilike("%{}%".format(search_term)),
```

Comment:
The only thing I'm wondering is that I think it's only useful for us to search the `to` field for letters, is that correct? As for emails and SMS it's taken care of by the search of `normalised_to`? 

I'm just wondering if this is going to have a performance impact for searching through SMS and emails as now it's also going to be doing an `ilike` on the `to` field which seems like it could be a fairly intensive query (and I don't know what indexing we have on that column to support that also)?

Do you think it might be an issue too? Ideally we could know by doing a before and after test what the impact would be and look at some stats (like https://grafana-paas.cloudapps.digital/d/u5LtjAqZk/notify-api-http?orgId=2) but it may just be worth trying to do some logic to only include the or statement for letters or `none`?

Sorry I didn't think about this in the previous PR review

Reply:
Some quick testing on the production database, running these queries 10 times:

```sql
select
  "to" from notifications
where
  service_id = '<GOV.UK Email>' and "normalised_to" like '%chris%'
limit 50;
```
Slowest query: 573ms
Fastest query: 32ms


```sql
select
  "to" from notifications
where
  service_id = '<GOV.UK Email>' and "to" ilike '%chris%'
limit 50;
```
Slowest query: 850ms
Fastest query: 28ms

***

I expect the fast queries are Postgres doing some kind of caching. The slowest queries for the `ilike` version don’t seem too much worse, so reckon we can live with them bring in there for a week until the `normalised_to` column has populated?

Reply:
You are probably right that we can deal with it for a week.

Also, I'm not sure if those queries are for the difference I was getting at. It's not whether the `like` on `normalised_to` is different in speed compared to the `ilike` on `to`, it's that how much slower is it to search with BOTH the `like` on `normalised_to` and `ilike` on `to` versus just the `like` on `normalised_to`, when in the case of emails and SMS, the `ilike` on `to` is of no benefit and is just adding query time.

Saying that, given you've tested this on GOV.UK email, I think most smaller services that use search would see little difference in response times.